### ISSUE TITLE
Investigate Possibility of Parallel Topic Builds

### ISSUE THREAD

**Issue by EWSoftware** [EWSoftware/SHFB#2]
*Moved from [CodePlex](https://shfb.codeplex.com/workitem/33859)*

Investigate the possibility of having BuildAssembler build topics in parallel.  That would require that all build
components in the stack are thread-safe though.  Possible changes and issues:
- Add a configuration property to enable parallel builds but default to the current synchronous build behavior.
- All build components would need an overriden virtual property to indicate if they support parallel execution.  The default if not overridden would be false.  Each build component and syntax generator would need checking to make sure it is thread-safe.  For components that contain nested components, they would need to check each nested component too.
- Concurrent updates to static members.  Use thread-safe constructs where needed.
- Use of non-thread-safe objects like the code colorizer.  Wrap them in ThreadLocal<T>?
- Copying of content files.  Track source and destination files to copy in thread-safe constructs.  Then, when the component is disposed of, copy the content.

**Comment by sharwell** [comment]
Would the topic of incremental builds be included in this topic, or separate?

**Comment by EWSoftware** [comment]
It's definitely a separate topic.  However, people have asked for incremental builds in the past but I don't see it as a straightforward option to implement and I haven't seen a viable way to do it.  In a very narrow scope considering conceptual topics only, perhaps it could be done in BuildAssembler since simplistically you'd only have to check to see if the source topic file had changed.

However, factor in that changing practically any project setting can affect the style or layout and it gets much more complicated, something as obvious as the presentation style or something less subtle such as a change to the header or footer text.  Content imported from snippet files, token files, external source code, and plug-ins and build components configuration changes could also change the topic in unpredictable ways that couldn't necessarily be detected.

Consider what would have to occur for API topics and I don't think it gets any easier.  BuildAssembler is usually the most time consuming step of the build  Trying to find ways to reduce time in steps prior to it through an incremental build process probably isn't worthwhile.  For now, I see the time better spent looking for ways to reduce the build time through other means.

**Comment by EWSoftware** [comment]
After experimenting with this, I've concluded it's not a worthwhile option.  All told, it only resulted in a few percent reduction in build time for most projects even the SHFB help files.  It's just not worth the effort to maintain thread safety in all of the build components for such a minimal return on the effort to support parallel builds.

Using a pipeline task to perform the writes for the IntelliSense and Save components gave a decent performance increase on systems with slow hard drives so I've retained those changes.  Combining the conceptual and reference builds into a single step also simplified things and gives a minor performance benefit as it doesn't have to load the common reference data in two separate steps anymore.
